src/Microsoft.DotNet.Build.Tasks/GatherFoldersToRestore.cs

```diff
@@ -0,0 +1,73 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Text;
+using System.IO;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+namespace Microsoft.DotNet.Build.Tasks
+{
+    public class GatherDirectoriesToRestore : Task
+    {
+        [Required]
+        public string[] RootDirectories { get; set; }
+
+        [Output]
+        public string[] DirectoriesToRestore { get; set; }
+
+        public int DirectoryListLength = 7000;
+
+        public override bool Execute()
+        {
+            HashSet<string> directoriesToRestore = new HashSet<string>();
+
+            foreach (string rootDirectory in RootDirectories)
+            {
+                AddDirectoriesToRestore(rootDirectory, rootDirectory, directoriesToRestore);
```

Comment:
This is a bit slower.  I think the major issue is when we pass in directories and let `dotnet restore` do the walking we only have to invoke `dotnet restore` more times (5 times vs 1 time).  There's a bit of overhead to actually boot the whole thing up and have it reload all of the state about the feeds, etc.  It's maybe ~15-20s slower in the case where there's no work to be done.  I'd rather keep what I propose.